server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,852 @@
+    linkdirs = sorted(hier.config.LINKDIRS.split(" "))
+    if directories:
+        for dirent in directories:
+            if all(
+                [
+                    os.path.exists(os.path.join(hier.path, controller, dirent)),
+                    dirent != "_QUARANTINED",
+                    not dirent.startswith("WONT-INDEX"),
+                ]
+            ):
+                if dirent not in linkdirs:
+                    hier.add_unexpected_dirs(controller, dirent)
+    else:
+        hier.set_status(controller, "subdirs")
+
+    return 0
+
+
+def verify_prefixes(hier, controller):
+    prefix_dir = os.path.join(hier.path, controller, ".prefix")
+    if not os.path.exists(prefix_dir):
+        return
+    if not os.path.isdir(prefix_dir):
+        hier.set_status(controller, "prefix_dir")
+        return
+
+    prefixes = glob.iglob(os.path.join(prefix_dir, "*"))
+
+    for prefix in prefixes:
+        base_prefix = os.path.basename(prefix)
+        if not base_prefix.startswith("prefix.") and not base_prefix.endswith(
+            ".prefix"
+        ):
+            hier.add_non_prefixes(controller, base_prefix)
+        elif base_prefix.startswith("prefix."):
+            hier.add_wrong_prefixes(controller, base_prefix)
+
+    return
+
+
+def verify_tarball_names(hier, controller):
+    pass
+
+
+def verify_archive(hier):
+    controllers = glob.iglob(os.path.join(hier.path, "*"))
+
+    for controller in controllers:
+        if os.path.isdir(controller):
+            hier.add_controller(os.path.basename(controller))
+        else:
+            hier.add_bad_controller(os.path.basename(controller))
+
+    # now check each "good" controller and get the tarballs it contains
+    for controller in hier.controllers:
+        direct_entries = glob.iglob(os.path.join(hier.path, controller, "*"))
+        hidden_entries = glob.glob(os.path.join(hier.path, controller, ".*"))
+        controller_subdir = list()
+        if hidden_entries:
+            for hid_entries in hidden_entries:
+                hier.add_unexpected_objects(controller, os.path.basename(hid_entries))
+        for item in direct_entries:
+            base_item = os.path.basename(item)
+            if os.path.isdir(item):
+                controller_subdir.append(base_item)
+            elif os.path.islink(item):
+                symlink_item = f"{base_item} -> {os.path.realpath(item)}"
+                hier.add_unexpected_symlinks(controller, symlink_item)
+            elif all(
+                [
+                    os.path.isfile(item),
+                    not base_item.endswith(".tar.xz"),
+                    not base_item.endswith(".tar.xz.md5"),
+                ]
+            ):
+                hier.add_unexpected_objects(controller, base_item)
+            elif os.path.isfile(item) and (
+                base_item.endswith(".tar.xz") or base_item.endswith(".tar.xz.md5")
+            ):
+                hier.add_tarballs(controller, base_item)
```

Comment:
What do we do with the tarballs?

Reply:
I have made it a LIST instead of a DICT so that it can just store controllers because at the end it is only checking for controllers that are not their in the controllers list.